---
title: "Exercise 4"
author: "Jonas Freimuth, 807218"
date: "17 11 2021"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

require("dplyr",      quietly = TRUE)
require("tidyr",      quietly = TRUE)
require("tidyselect", quietly = TRUE)
require("pegas",      quietly = TRUE)

# disable grouping output message
options(dplyr.summarise.inform = FALSE)

```

## Task 1

```{r task_1_1}
loci_dat <- VCFloci("data/Ex4.vcf", quiet = TRUE)

n_loci <- nrow(loci_dat)
```

```{r task_1_2}
n_snps_is <- sum(is.snp(loci_dat))


vi_inf <- getINFO(loci_dat, "VT")
n_snps_gi <- sum(vi_inf == "SNP")

cat("Variant types:\n")
cat(paste(unique(vi_inf), collapse = ", "), "\n")
```

\# of SNPs from `is.snp`: `r n_snps_is`  
\# of SNPs from `getINFO`: `r n_snps_gi`  

$\Rightarrow$ `is.snp` gives `r abs(n_snps_is - n_snps_gi)` less than `getINFO`, 
due to the former not counting one locus recorded in `VT` as a SNP.

## Task 2

```{r task_2_1}
vc_dat <- read.vcf("data/Ex4.vcf", quiet = TRUE)

loc_1 <- as.loci(vc_dat[[1]])

loc_1_freq <- summary(loc_1)
loc_1_freq

all_freq <- loc_1_freq$x$allele   / sum(loc_1_freq$x$allele)
gen_freq <- loc_1_freq$x$genotype / sum(loc_1_freq$x$genotype)
 
maj_all <- all_freq[all_freq == max(all_freq)]
min_all <- all_freq[all_freq == max(all_freq)]
```

$\Rightarrow$ * Major allele frequency: $f_{`r names(maj_all)`}$
$= `r round(maj_all, 3)`$  
* Minor allele frequency: $f_{`r names(min_all)`}$
$= `r round(min_all, 3)`$  

## Task 3

```{r task_3_1}
loc_1_hw <- hw.test(loc_1)
loc_1_hw
```
$\Rightarrow$ The observed genotype frequencies are different from the ones 
expected under Hardy-Weinberg equilibrium at a high significance level.  
$\Rightarrow$ $\chi^2$-statistic $\rightarrow$ $\chi^2$-test is performed. (Also
 a Monte Carlo permutation test is performed, which leads to the `Pr.exact` 
 result.)
 
```{r task_3_2}
# expected genotype frequencies
exp_gen_freq <- c("C/C" = min_all           ^ 2,
                  "C/T" = min_all * maj_all * 2,
                  "T/T" = maj_all           ^ 2)

# expected genotype numbers
exp_gen_num  <- exp_gen_freq * nrow(loc_1)

# observed genotype numbers
obs_gen_num  <- gen_freq     * nrow(loc_1)

# calculate chi_sq statistic
chi_sq <- sum(((obs_gen_num - exp_gen_num) ^ 2) / exp_gen_num)

# get pval from chi_sq statistic
pval_hw_loc_1 <- pchisq(chi_sq, 1, lower.tail = FALSE)

pval_hw_loc_1
```
 $\Rightarrow$ My own method of calculation agrees with the `pegas` function.
 
## Task 4

```{r task_4_1}
loci_11_20 <- as.loci(vc_dat[, 11:20])

ld_score_mat <- as.matrix(LDscan(loci_11_20, quiet = TRUE))

opar <- par(oma = rep(0.5, 4))
LDmap(ld_score_mat, asp = 1)
```

```{r task_4_2}
max_ld_score <- which(ld_score_mat == max(ld_score_mat), arr.ind = TRUE)
max_ld_pair  <- row.names(max_ld_score)
```

$\Rightarrow$ The highest scoring SNP pair is locus `r max_ld_pair[1]` and locus
`r max_ld_pair[2]`, with $D_{adj} = `r max(ld_score_mat)`$

```{r task_4_3}
# extract highest score loci
loc_ab <- as.loci(loci_11_20[ , max_ld_score[1, ]])

# extract allele freqs
loc_ab_freq <- summary(loc_ab)

ab_all_freq <- sapply(loc_ab_freq, function(x) {
   x$allele / sum(x$allele)
})

# find major allels
ab_maj_alls <- sapply(loc_ab_freq, function(x) {
   names(x$allele[x$allele == max(x$allele)])
})

# get genotype freqs of individuals
hap <- haplotype(loc_ab, quiet = TRUE)

# scale to get relative freqs
ab_gen_freq <- (attr(hap, "freq") / sum(attr(hap, "freq")))

# get index of major major genotype
AB_ind <- which(hap[1, ] == ab_maj_alls[1] & hap[2, ] == ab_maj_alls[2])

# get rel freq of major major genotype
p_AB <- ab_gen_freq[AB_ind]

# get indices of major allele freqs
A_ind <- which(ab_all_freq == max(ab_all_freq[, 1]), arr.ind = TRUE)
B_ind <- which(ab_all_freq == max(ab_all_freq[, 2]), arr.ind = TRUE)

# get expected rel freq of major major genotype
p_A.p_B <- ab_all_freq[A_ind] * ab_all_freq[B_ind]

# calculate disequilirium score and adjusted score
D <- p_AB - p_A.p_B
D_adj <- D / sqrt(prod(ab_all_freq))

# display adjusted score
D_adj
```

$\Rightarrow$ The manual method returns the same result as the `pegas` function.